synapse/module_api/__init__.py

```diff
@@ -141,7 +141,7 @@
 
 PRESENCE_ALL_USERS = PresenceRouter.ALL_USERS
 
-ALLOW = spam_checker_api.Allow.ALLOW
+ALLOW = Allow.ALLOW
```

Comment:
Why are we defining ALLOW here but not exporting it via `__all__`?

For that matter, why are we re-exporting `Allow`?

Also, `module_api.ALLOW` doesn't seem like a great name to me. what happens if we have more than one thing that has an ALLOW/DENY interface, and uses different constants?

Reply:
> Why are we defining ALLOW here but not exporting it via `__all__`?

Actually I misread the docs and it looks like it's ALLOW we need to export, not Allow.

> what happens if we have more than one thing that has an ALLOW/DENY interface, and uses different constants?

I'm not sure I see the issue. We currently don't have that on user-facing interfaces (apart from other callbacks returning True/False, which could arguably just be made to return `ALLOW` and another value we'd introduce for deny), and if we get more in the future we should just make them use that?

Reply:
It just feels like a rather opaque name for something related to the spam checker. If I look at the exports for `module_api`, I see a thing called ALLOW and wonder what it is meant to do.

> I'm not sure I see the issue. We currently don't have that on user-facing interfaces (apart from other callbacks returning True/False, which could arguably just be made to return ALLOW and another value we'd introduce for deny), and if we get more in the future we should just make them use that?

Well, for example, let's suppose we decided to apply the same logic to `RegistrationBehaviour`. Now we need to export things called `ALLOW`, `SHADOW_BAN` and `DENY` through the `module_api`, and it's a *different* `ALLOW` to the existing `ALLOW`.

Anyway, maybe it's fine. I just think we ought to try to make sure that this module doesn't end up with millions of exports with unclear purposes.

Reply:
Right, after discussing this in #synapse-dev, we have come to the conclusion that the right course of action is replacing `Allow` with a single constant that is `NOT_SPAM`, which makes things clearer and less confusing, and doesn't use an enum to store just one value.

Reply:
If this is the most Pythonesque way to do it, then sure.
However, note that this breaks the `str` return value of `check_event_for_spam`, since we now have one magical `str` that means "pass" and other `str` values that mean "fail".

Reply:
> As far as I can tell, this breaks the str return value of check_event_for_spam, since we now have one magical str that means "pass" and every other str value means "fail".

That was the case before, though. `ALLOW` was a magical str, with the value `allow`. It has some extra metadata, but fundamentally it's still just a magical `str`.

If you really want to do this in a type-enforced manner, the Pythonesque way to create a special sentinel value (other than `None`) is to create an object, rather than using an enum with a single entry:

```
NOT_SPAM = object()
```

Personally though, it's not the use of magical values for strings that bothered me here. It was the use of a completely generic name which didn't give any clues about which interfaces it was supposed to be used with.